File: apps/mobile/test/widgets/friend_button_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:mobile/core/theme.dart';
import 'package:mobile/models/friendship.dart';
import 'package:mobile/widgets/friend_button.dart';

import '../support/fake_api_client.dart';

void main() {
  testWidgets(
    'NONE : envoyer une demande passe le bouton sur "Demande envoyée"',
    (tester) async {
      final api = FakeApiClient()
        ..onSendFriendRequest = (pseudo) async => FriendshipSummary(
          id: 'f1',
          status: FriendshipStatus.pending,
          createdAt: DateTime(2026, 1, 1),
          user: FriendUserSummary(pseudo: pseudo, avatarUrl: null),
        );

      await tester.pumpWidget(
        MaterialApp(
        theme: buildReverbLightTheme(),
          home: Scaffold(
            body: FriendButton(
              api: api,
              pseudo: 'bob',
              initialStatus: ViewerFriendshipStatus.none,
              initialFriendshipId: null,
            ),
          ),
        ),
      );

      await tester.tap(find.text('Ajouter en ami'));
      await tester.pumpAndSettle();

      expect(find.text('Demande envoyée'), findsOneWidget);
    },
  );

  testWidgets('PENDING_RECEIVED : accepter passe le bouton sur "Ami·e"', (
    tester,
  ) async {
    final api = FakeApiClient()..onAcceptFriendRequest = (id) async {};

    await tester.pumpWidget(
      MaterialApp(
        theme: buildReverbLightTheme(),
        home: Scaffold(
          body: FriendButton(
            api: api,
            pseudo: 'bob',
            initialStatus: ViewerFriendshipStatus.pendingReceived,
            initialFriendshipId: 'f1',
          ),
        ),
      ),
    );

    await tester.tap(find.text('Accepter'));
    await tester.pumpAndSettle();

    expect(find.textContaining('Ami'), findsOneWidget);
  });

  testWidgets('SELF ne rend rien', (tester) async {
    final api = FakeApiClient();

    await tester.pumpWidget(
      MaterialApp(
        theme: buildReverbLightTheme(),
        home: Scaffold(
          body: FriendButton(
            api: api,
            pseudo: 'moi',
            initialStatus: ViewerFriendshipStatus.self,
            initialFriendshipId: null,
          ),
        ),
      ),
    );

    expect(find.byType(FriendButton), findsOneWidget);
    expect(find.text('Ajouter en ami'), findsNothing);
  });
}

File: apps/web/src/lib/components/profile/FriendButton.test.ts
import { describe, expect, it, vi } from 'vitest';
import { render, screen, fireEvent, waitFor } from '@testing-library/svelte';
import FriendButton from './FriendButton.svelte';

const { sendFriendRequest, acceptFriendRequest } = vi.hoisted(() => ({
	sendFriendRequest: vi.fn(),
	acceptFriendRequest: vi.fn()
}));

vi.mock('$lib/api/client', async (importOriginal) => {
	const actual = await importOriginal<typeof import('$lib/api/client')>();
	return { ...actual, api: { ...actual.api, sendFriendRequest, acceptFriendRequest } };
});

describe('FriendButton', () => {
	it('NONE : envoyer une demande passe le bouton sur "Demande envoyée"', async () => {
		sendFriendRequest.mockResolvedValue({ id: 'f1', status: 'PENDING' });
		render(FriendButton, { pseudo: 'bob', status: 'NONE', friendshipId: null });

		await fireEvent.click(screen.getByRole('button', { name: 'Ajouter en ami' }));

		await waitFor(() =>
			expect(screen.getByRole('button', { name: 'Demande envoyée' })).toBeInTheDocument()
		);
	});

	it('PENDING_RECEIVED : accepter passe le bouton sur un état ami', async () => {
		acceptFriendRequest.mockResolvedValue(undefined);
		render(FriendButton, { pseudo: 'bob', status: 'PENDING_RECEIVED', friendshipId: 'f1' });

		await fireEvent.click(screen.getByRole('button', { name: 'Accepter' }));

		await waitFor(() => expect(acceptFriendRequest).toHaveBeenCalledWith('f1'));
		expect(screen.getByText(/Ami/)).toBeInTheDocument();
	});

	it('SELF ne rend rien', () => {
		render(FriendButton, { pseudo: 'moi', status: 'SELF', friendshipId: null });

		expect(screen.queryByText('Ajouter en ami')).not.toBeInTheDocument();
	});
});
